WIP: [Refactor] Training Engine Interface and Development Plan

ZihengJiang opened the pull request:
# [Refactor] Training Engine Interface and Development Plan

## Motivation  
See the original RFC for background: https://github.com/volcengine/verl/issues/1371  

Modernizing our training loop requires that we:

- **Decouple** training-backend implementation from algorithm code so each can evolve independently  
- **Unify** on a single, well-defined `Engine` interface across FSDP/Megatron/etc backends  
- **Enable** unit-testing of each backend implementation in isolation  
- **Guarantee** algorithm “roles” (Critic, Actor, Rollout, Ref) remain completely engine-agnostic.
- **Preserve** the existing Roles interface exactly—so prior algorithms (e.g., `PPOTraining`) continue to work without change 

---

## Current Implementation  


### Classic Training Loop with the New Interface

```python
# 1. Build and initialize engine
engine = FSDPEngine(config)
engine.init_model()
engine.set_loss_fn(loss_fn)

# 2. Training loop
for epoch in range(config.num_epochs):
    for batch in train_loader:
        # a) zero gradients
        engine.optimizer_zero_grad()

        # b) forward + backward
        with engine.train_mode():
            preds, loss, ctx = engine.forward_backward_step(
                batch,
                ctx,
                forward_only=False,
                preprocess_fn=preprocess_fn,
                postprocess_fn=postprocess_fn
            )

        # c) update and schedule
        grad_norm = engine.optimizer_step()
        current_lr = engine.lr_scheduler_step()

# 3. Evaluation
with engine.eval_mode():
    for micro_batch in data:
        preds, ctx = engine.forward_backward_step(
            micro_batch,
            ctx,
            forward_only=True,
            preprocess_fn=preprocess_fn,
            postprocess_fn=postprocess_fn
        )
```

### Detailed BaseEngine Interface
We now introduce an abstract base class, `BaseEngine`, exposing core methods:

```python

class BaseEngine(object):
    """
    Abstract base class defining the interface for model training engines.

    Engine implementations must subclass BaseEngine and provide concrete behavior for all methods.
    """
    def __init__(self, config):
        """
        Initialize the BaseEngine.

        Args:
            config: Configuration object containing parameters for engine setup.
        """
        raise NotImplementedError

    def init_model(self):
        """
        Instantiate or load the model, optimizer, and learning rate scheduler.

        Should prepare all components necessary for training or evaluation.
        """
        raise NotImplementedError

    def train_mode(self):
        """
        Context manager entry for switching the engine and model into training mode.

        Usage:
            with engine.train_mode():
                # runs in training mode
        """
        raise NotImplementedError

    def eval_mode(self):        
        """
        Context manager entry for switching the engine and model into evaluation mode.

        Usage:
            with engine.eval_mode():
                # runs in evaluation mode
        """
        raise NotImplementedError

    def forward_backward_step(self, 
                              batch, 
                              ctx=None, 
                              forward_only=False, 
                              preprocess_fn=None, 
                              postprocess_fn=None):
        """
        Execute a forward pass (and optional backward pass) over a batch of data.

        Args:
            batch: Raw batch data (e.g., tensors or mappings) to process.
            ctx: Optional context dict passed to preprocess/postprocess functions.
            forward_only: If True, skip gradient computation and backward pass.
            preprocess_fn: Function(batch, ctx) -> (inputs, ctx), applied before model call.
            postprocess_fn: Function(outputs, ctx) -> (predictions, ctx), applied after model call.

        Returns:
            If forward_only:
                (predictions, ctx)
            Else:
                (predictions, loss, ctx)
        """
        raise NotImplementedError

    def optimizer_zero_grad(self):
        """
        Zero out gradients of all parameters before starting a new backward pass.
        """
        raise NotImplementedError

    def optimizer_step(self):
        """
        Perform an optimization step to update model parameters based on accumulated gradients.

        Returns:
            grad_norm (float): The norm of the gradients before clipping or update.
        """
        raise NotImplementedError

    def lr_scheduler_step(self):
        """
        Advance the learning rate scheduler by one step.

        Returns:
            current_lr (float or list[float]): Updated learning rate(s).
        """
        raise NotImplementedError

    def shard_data(self, data):
        """
        Shard or partition data for distributed training or parallel execution.

        Args:
            data: Data structure to be sharded across devices/workers.

        Returns:
            Sharded data in the same format as input.
        """
        raise NotImplementedError

    def unshard_data(self, data):
        """
        Reconstruct or gather sharded data back to a unified format.

        Args:
            data: Sharded data structure to reconstruct.

        Returns:
            Unsharded, combined data.
        """
        raise NotImplementedError
        

    def set_loss_fn(self, loss_fn):
        """
        Set the loss function to be used during training.

        Args:
            loss_fn: Callable(data, predictions, ctx) -> (loss_tensor, new_ctx)
        """
        raise NotImplementedError

    def to(self, device: str, model: bool = True, optimizer: bool = True):
        """
        Move model parameters, optimizer states, or both to the specified device.

        Args:
            device: Target device identifier (e.g., "cuda" or "cpu").
            model: If True, move the model.
            optimizer: If True, move the optimizer states.
        """
        raise NotImplementedError


    def save_checkpoint(self, local_path, hdfs_path=None, global_step=0, max_ckpt_to_keep=None):
        """
        Save model, optimizer, and scheduler states to a checkpoint.

        Args:
            local_path: Local filesystem path to save checkpoint.
            hdfs_path: Optional HDFS path to copy checkpoint.
            global_step: Integer training step number for naming.
            max_ckpt_to_keep: Maximum number of recent checkpoints to retain.
        """
        raise NotImplementedError


    def load_checkpoint(self, local_path, hdfs_path=None, del_local_after_load=True):
        """
        Load model, optimizer, and scheduler states from a checkpoint.

        Args:
            local_path: Local filesystem path of the checkpoint.
            hdfs_path: Optional HDFS path where checkpoint is stored.
            del_local_after_load: Whether to delete local copy after loading.
        """
        raise NotImplementedError
```

### FSDPEngine Implementaion

A concrete `FSDPEngine` implements all methods using PyTorch FullyShardedDataParallel, supporting all the features that FSDP DPCritic Worker support:

- Multi-GPU/model sharding  
- Activation- and optimizer-offload  
- LoRA & sequence parallelism  
- Dynamic batch size and remove padding

### CriticWorker Implementation based on the FSDPEngine
- Unchanged public API 
- Each role calls only BaseEngine methods (init_model, train_mode/eval_mode, forward_backward_step, etc.)
- No modifications needed in existing algorithms (e.g., PPOTraining)
- New roles can be plugged in identically to legacy code

## Development Plan
We’ll roll this out in three gated phases, controlled by a feature-flag (`use_legacy_worker_impl`).

### Phase 1: Engine Development
> Flag: use_legacy_worker_impl = True (default)
> New interface under active development

- Refactor Critic, Actor, Rollout, Ref to use only BaseEngine APIs
- Design a hierarchical, immutable config system for engine/backends
- Ensure PPO training curves and final accuracy match legacy implementation

### Phase 2: Migration
> Flag: use_legacy_worker_impl = False (default) – legacy path logs a deprecation warning
> All new code targets the new interface; 2–3 months of integration/stress testing

- Enforce new interface for all feature work
- Gather benchmarks, bug reports, and performance data

### Phase 3: Cleanup
> After Phase 2 validation:
- Remove legacy worker code and flags
- Delete deprecated paths
- Finalize documentation, update changelogs, close deprecation notices

Please review this refactor and share any feedback or concerns!

ZihengJiang commented:
@vermouth1992 @eric-haibin-lin @PeterSH6 @tongyx361 @ETOgaosion @hongpeng-guo @wwwjn @tianyu-l
Please review this refactor and share any feedback or concerns!

ZihengJiang commented:
> > using PyTorch FullyShardedDataParallel.
> 
> Meaning FSDP1?

I also migrated the FSDP2 implementation in the PR. https://github.com/volcengine/verl/blob/a7459481a43596e942102b878012672e900612d5/verl/workers/engine/fsdp/engine_impl.py#L248

ISEEKYAN commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,362 @@
+
+                with torch.no_grad():
+                    # TODO: should not access module in the engine
+                    use_value_head_model = hasattr(self.engine.module, "v_head")
```
this is not used in megatron

ISEEKYAN commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,362 @@
+
+        # normalize config
+        self.config.ppo_mini_batch_size *= self.config.rollout_n
+        self.config.ppo_mini_batch_size //= torch.distributed.get_world_size() // self.ulysses_sequence_parallel_size
```
ulysses_sequence_parallel_size is a concept of FSDP only, megatron do not use it

vermouth1992 replied:
agreed. ulysses_sequence_parallel_size should be moved into FSDPEngine

ISEEKYAN commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,362 @@
+        self.engine = FSDPEngine(self.config)
+
+        def loss_fn(batch, vpreds, ctx):
+            responses = batch["responses"]
+            attention_mask = batch["attention_mask"]
+            values = batch["values"]
+            returns = batch["returns"]
```
do we have a protocol for batch?

ISEEKYAN commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,362 @@
+
+                        self.engine.optimizer_zero_grad()
+
+                        for micro_batch in micro_batches:
```
the iteration through mini-batch is wrapped by callbacks in megatron, instead of directly iterating every micro batch and calling a `forward_backward_step`. 
In megatron it is more proper to call a function `forward_backward_mini_batch`.

ISEEKYAN reviewed (commented):
Hello, I have some comments from aspect of megatron compatibility.

vermouth1992 commented on verl/workers/engine/fsdp/engine_impl.py:
```diff
@@ -0,0 +1,541 @@
+        return self.ulysses_sharding_manager.postprocess_data(data)
+
+
+    def forward_backward_step(self, 
```
forward_backward_step should train a batch, run forward + backward + optimizer step

ZihengJiang commented:
@vermouth1992 @ISEEKYAN made several updates:
- adapt the forward-backward-step interface for mini-batch instead of microbatch
- move ulysses config into engine implementation and left microbatch process logic also to the engine.

Let me know whether current interface looks good to you guys.

vermouth1992 commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,288 @@
+    def init_model(self):
+        self.engine.init_model()
+
+    def get_microbatch_process_fn(self):
```
Also, make the megatron to "fill in the black" for this function as well

vermouth1992 replied:
Shall we move those preprocess/postprocess function into a separate file as I expect it to be very large

vermouth1992 replied:
Maybe we can make them a class so that other backends can subclass directly

ZihengJiang replied:
use class is a good idea

vermouth1992 commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,288 @@
+        output = output.to("cpu")
+        return output
+
+    @register(dispatch_mode=Dispatch.DP_COMPUTE_PROTO)
```
We will need to decide how we dispatch here.

ZihengJiang replied:
can you elaborate on this?

eric-haibin-lin replied:
we'll work on that for megatron backend

vermouth1992 commented on verl/workers/engine/fsdp/engine_impl.py:
```diff
@@ -0,0 +1,667 @@
+            revert_indices = torch.tensor(get_reverse_idx(indices), dtype=torch.long)
+            values = values[revert_indices]
+
+        response_mask = data.batch["response_mask"]
```
We should move those logics to worker as they are critic specific.

eric-haibin-lin commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,288 @@
+        self.config = config
+        self.engine = FSDPEngine(self.config)
+
+        def loss_fn(batch, vpreds, ctx):
```
can we avoid nested function and make it a member function and pass self.loss_fn to engine? We should make `__init__` more modular and also make it easier for others to override the loss fn in their extensions
```
def loss_fn(self, batch, vpreds, ctx)
```

eric-haibin-lin commented on verl/workers/roles/critic.py:
```diff
@@ -0,0 +1,288 @@
+logger = logging.getLogger(__file__)
+logger.setLevel(os.getenv("VERL_LOGGING_LEVEL", "WARN"))
+
+
```
can we add some unit test for worker and engine in this PR?

eric-haibin-lin reviewed (commented):
nice!

ZihengJiang commented:
Hi @vermouth1992 @eric-haibin-lin 

I've made several updates to the PR and addressed most of the comments. The infer_batch and train_batch interfaces have gone through a few iterations and now support multi-output and callback hooks, should making them flexible enough for various worker implementations.

Could you let me know the preferred approach for testing? Do we have a common used e2e test workload?

eric-haibin-lin commented on verl/trainer/config/ppo_trainer.yaml:
```diff
@@ -925,6 +925,9 @@ trainer:
   device: cuda
 
+  # whether to use legacy role implementation
+  use_legacy_worker_impl: True
```
can we default it to null or "auto", so that we can automatically switch to the new impl if user do not set this config
